src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/JpaTimelinePersistenceAdapter.kt

```diff
@@ -3,33 +3,68 @@ package uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa
 import mu.KotlinLogging
 import org.springframework.stereotype.Component
 import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.entity.TimelineEntity
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.entity.TimelineEntity.Companion.newTimelineForPrisoner
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.mapper.TimelineEntityMapper
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.mapper.TimelineEventEntityMapper
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.app.database.jpa.repository.TimelineRepository
+import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.timeline.Timeline
 import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.timeline.TimelineEvent
 import uk.gov.justice.digital.hmpps.educationandworkplanapi.domain.timeline.service.TimelinePersistenceAdapter
 
 private val log = KotlinLogging.logger {}
 
 @Component
 class JpaTimelinePersistenceAdapter(
-  // private val timelineRepository: TimelineRepository,
-  // private val timelineMapper: TimelineEntityMapper,
+  private val timelineRepository: TimelineRepository,
+  private val timelineMapper: TimelineEntityMapper,
+  private val timelineEventMapper: TimelineEventEntityMapper,
 ) : TimelinePersistenceAdapter {
 
   @Transactional
-  override fun recordTimelineEvent(prisonNumber: String, event: TimelineEvent) {
+  override fun recordTimelineEvent(prisonNumber: String, event: TimelineEvent): TimelineEvent {
     log.info { "Recording TimelineEvent [$event] for prisoner [$prisonNumber]" }
-    // TODO RR-317
+
+    val timelineEntity = findOrCreateTimelineEntity(prisonNumber)
+    val timelineEventEntity = timelineEventMapper.fromDomainToEntity(event)
+
+    with(timelineEntity) {
+      addEvent(timelineEventEntity)
+      timelineRepository.saveAndFlush(this)
+    }
+
+    // use the persisted entity with the populated JPA fields, rather than the non persisted entity reference above
+    val persisted = timelineEntity.events!!.first { it.reference == timelineEventEntity.reference }
+    return timelineEventMapper.fromEntityToDomain(persisted)
   }
 
   @Transactional
-  override fun recordTimelineEvents(prisonNumber: String, events: List<TimelineEvent>) {
+  override fun recordTimelineEvents(prisonNumber: String, events: List<TimelineEvent>): Timeline {
     log.info { "Recording [${events.size}] TimelineEvents for prisoner [$prisonNumber]" }
-    // TODO RR-317
+
+    val timelineEntity = findOrCreateTimelineEntity(prisonNumber)
+    events.forEach {
```

Comment:
This further addresses [a comment](https://github.com/ministryofjustice/hmpps-education-and-work-plan-api/pull/91#discussion_r1324201100) @nathanrussell-moj-digital made on my previous PR re where this loop lives.